### ISSUE TITLE
Many can't be sure that Hyprland/Xwayland crash issues may be related to file descriptors

### ISSUE THREAD

**Issue by hdm9527** [hyprwm/Hyprland#6584]
### Bug or Regression?

Bug

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch makepkg at commit d1340bd1d8eedd274283e0cb2568a3ed67b58c81  (keybinds: ignore missing keysyms if no other methods match).
Date: Mon Jun 17 23:53:51 2024
Tag: v0.41.1-40-gd1340bd1, commits: 4858

flags: (if any)
```

</details>

### Description

Hyprland crashed just now. I checked coredumpctl and found that fd failed to be set, but then I looked at the source code and found that it was a call to the select() method that caused fd to run out, remembered a [blog](https://0pointer.net/blog/file-descriptor-limits.html) I read recently that using select() might randomly crash. https://news.ycombinator.com/item?id=27216242 Here's what some people are saying about select(). Related issues #6398 #5321 #6555

![image](https://github.com/hyprwm/Hyprland/assets/39702328/f05b3606-baaf-4683-a466-822df232a3a5)
![image](https://github.com/hyprwm/Hyprland/assets/39702328/90ff3c67-0df8-47b3-bc2c-2f39dbb1b258)

### How to reproduce

above

**Comment by vaxerski** [comment]
Makes sense, thanks for the insights.

**Comment by vaxerski** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/43317083/073304b8-f84b-47d2-b938-e2f1f696f635)
Maybe this is why I never had any issues, lol.

**Comment by vaxerski** [comment]
I've pushed two commits regarding this, one rids of `select` and the other bumps nofile limits as much as possible. Please check now.

**Comment by hdm9527** [comment]
b98e087 Using epoll() instead of poll() might be better, actual performance needs to be tested. Note that epoll() is linux only, FreeBSD needs kqueue() and NetBSD needs poll(). This [blog](https://hechao.li/2022/01/04/select-vs-poll-vs-epoll/) talks about the difference between select, poll and epoll.
https://gitlab.freedesktop.org/wayland/wayland/-/issues/72 https://gitlab.freedesktop.org/wayland/wayland/-/issues/214

**Comment by hdm9527** [comment]
RLIMIT_NOFILE is more elegantly set up in this way https://github.com/swaywm/sway/commit/38020d157ddb58e756c654e9a2ff203c1562b25b

**Comment by hdm9527** [comment]
> ![image](https://private-user-images.githubusercontent.com/43317083/340811728-073304b8-f84b-47d2-b938-e2f1f696f635.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NzkwMTYsIm5iZiI6MTcxODc3ODcxNiwicGF0aCI6Ii80MzMxNzA4My8zNDA4MTE3MjgtMDczMzA0YjgtZjg0Yi00N2QyLWI5MzgtZTJmMWY2OTZmNjM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDA2MzE1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlOTgyYmU5ZTY5MzNjNWY4YWY4MWIyYmIyMjVjZjc3MWZhZjNhYjNhMjgwMDg4MWU1NWQ2N2I0MGNhMzY5NTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ef0tOkMje2VkHywQrdGP3cpeSeleSWFrvklxJ9t6fRk) 
Maybe this is why I never had any issues, lol.

The problem should be this https://github.com/hyprwm/Hyprland/issues/6398#issuecomment-2175532044 https://github.com/swaywm/wlroots/issues/2594

**Comment by vaxerski** [comment]
thanks, I'll adapt the patch. Gimme a moment.

**Comment by vaxerski** [comment]
there

**Comment by hdm9527** [comment]
Two questions, I think it is possible to epoll as it gives better performance. Note the reuse of wl_buffers, yesterday I noticed that the dmabuf has been increasing without decreasing,fd has about 900 is dmabuf, running application did not increase!

```
❯ ls /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
897
```

**Comment by hdm9527** [comment]
> oopsie, looks like a fd leak somewhere
> 
> doesn't happen on my end though:
> 
> ![image](https://private-user-images.githubusercontent.com/43317083/341206916-d9fdc145-b55e-418c-a401-94e6f207b17d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4NDg0MjcsIm5iZiI6MTcxODg0ODEyNywicGF0aCI6Ii80MzMxNzA4My8zNDEyMDY5MTYtZDlmZGMxNDUtYjU1ZS00MThjLWE0MDEtOTRlNmYyMDdiMTdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDAxNDg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxNjUzYTMwODNjYmQ0OWEwYzZkODE0ODYyYjAzZTVmYjgxZTIxOGYxYjI1OTk0ZjQ5Mzg2YWJiMmQ5MzY5MTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5hNdMPX0cHzi0xHatfXrze-sLeroS0kDXpWOHFhtBO4)
> 
> any repro case?

It seems to be an issue with the xwayland app. e.g. intellij, chromium(xwl)

Before intellij startup
```
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
51
```
intellij running
```
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
65
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
63
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
63
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
63
```
After intellij is closed
```
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
61
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
61
```

**Comment by hdm9527** [comment]
3 hours in Hyprland running intellij
```sh
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
401
```

**Comment by vaxerski** [comment]
do they go away when the app is closed? huh.

**Comment by hdm9527** [comment]
It's normal under wayland, xwayland creates far more fd's than it destroys.

**Comment by firedes** [comment]
Is this high memory usage of Hyprland related to this issue?
![image](https://github.com/hyprwm/Hyprland/assets/16650765/2fc0c0d0-af37-45fc-94ab-14c6260b2545)

**Comment by hdm9527** [comment]
> Is this high memory usage of Hyprland related to this issue? ![image](https://private-user-images.githubusercontent.com/16650765/341663739-2fc0c0d0-af37-45fc-94ab-14c6260b2545.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NTg4NTEsIm5iZiI6MTcxODk1ODU1MSwicGF0aCI6Ii8xNjY1MDc2NS8zNDE2NjM3MzktMmZjMGMwZDAtYWYzNy00NWZjLTk0YWItMTRjNjI2MGIyNTQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDA4MjkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhZWI2NDZmODkwYmViOTVmMzUxMDRiOGIzNWQwNzdhYTcwZjQxZDY1NDlmNzgxZDQ1ZTg1NTA1Y2FjZTMzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4uLj-HxscfODRf4vHrIPGjZAtHGnH0AlRWOraVjVYXo)

version？

**Comment by firedes** [comment]
> > Is this high memory usage of Hyprland related to this issue? ![image](https://private-user-images.githubusercontent.com/16650765/341663739-2fc0c0d0-af37-45fc-94ab-14c6260b2545.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NTg4NTEsIm5iZiI6MTcxODk1ODU1MSwicGF0aCI6Ii8xNjY1MDc2NS8zNDE2NjM3MzktMmZjMGMwZDAtYWYzNy00NWZjLTk0YWItMTRjNjI2MGIyNTQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDA4MjkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhZWI2NDZmODkwYmViOTVmMzUxMDRiOGIzNWQwNzdhYTcwZjQxZDY1NDlmNzgxZDQ1ZTg1NTA1Y2FjZTMzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4uLj-HxscfODRf4vHrIPGjZAtHGnH0AlRWOraVjVYXo)
> 
> version？

hyprland-git 0.41.1.r43.e0e3c4c6-1

**Comment by hdm9527** [comment]
> > > Is this high memory usage of Hyprland related to this issue? ![image](https://private-user-images.githubusercontent.com/16650765/341663739-2fc0c0d0-af37-45fc-94ab-14c6260b2545.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NTg4NTEsIm5iZiI6MTcxODk1ODU1MSwicGF0aCI6Ii8xNjY1MDc2NS8zNDE2NjM3MzktMmZjMGMwZDAtYWYzNy00NWZjLTk0YWItMTRjNjI2MGIyNTQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDA4MjkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhZWI2NDZmODkwYmViOTVmMzUxMDRiOGIzNWQwNzdhYTcwZjQxZDY1NDlmNzgxZDQ1ZTg1NTA1Y2FjZTMzMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4uLj-HxscfODRf4vHrIPGjZAtHGnH0AlRWOraVjVYXo)
> > 
> > 
> > version？
> 
> hyprland-git 0.41.1.r43.e0e3c4c6-1
> 
> Now 8G： ![image](https://private-user-images.githubusercontent.com/16650765/341710521-cb2d3fd1-c385-4588-ac2e-ed098f817cf6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NjE1NzgsIm5iZiI6MTcxODk2MTI3OCwicGF0aCI6Ii8xNjY1MDc2NS8zNDE3MTA1MjEtY2IyZDNmZDEtYzM4NS00NTg4LWFjMmUtZWQwOThmODE3Y2Y2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDA5MTQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY3Njk5MDA4MmRkZmM5MDVhYjU4MzllMTE4NzZiNzc2MjJhNzM4ODM2ODk1ZDBlMjI3ZDlmMWM2YzQ4ZWNkYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WAk0MprD2JevH-lCmXxbKgBaSEGKTo_RuINL6oElSQU) ![image](https://private-user-images.githubusercontent.com/16650765/341710680-c97780ee-f9d2-42cd-9e6a-da94c360892f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NjE1NzgsIm5iZiI6MTcxODk2MTI3OCwicGF0aCI6Ii8xNjY1MDc2NS8zNDE3MTA2ODAtYzk3NzgwZWUtZjlkMi00MmNkLTllNmEtZGE5NGMzNjA4OTJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDA5MTQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgyYzYxMjU4ZjA4OTlhZTNkNWZhZWJmOTMwZWEwOTEyMjk3NDk4MzNlZDE1YjA3ZTQyNjY4NmE5YjEzNmM4YWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.L4qhzc25GamE1MjXpl9_-RRvrCJOk4yB7qYpMEtVtvA)

`cat /proc/$(pidof Hyprland)/status | grep VmRSS`

**Comment by firedes** [comment]
☁  ~  grep VmRSS /proc/$(pidof Hyprland)/status
VmRSS:	9165716 kB

**Comment by hdm9527** [comment]
> ☁ ~ grep VmRSS /proc/$(pidof Hyprland)/status VmRSS: 9165716 kB

It may be related to the fd leak. Can you reproduce it?

**Comment by vaxerski** [comment]
likely a shared ptr to a `CDMABuffer` is sticking around somewhere then. or, well, many of them. I use some xwayland apps (very few) and havent seen this at all. Maybe someone can find why the buffers stick around. Relevant protocol would be protocols/linuxdmabuf

**Comment by sifmelcara** [comment]
Experienced same dmabuf leak issue with Xwayland Qt app (krita), it leaks ~500 dmabuf per hour and crashes Hyprland/Xwayland. The workaround mentioned in this issue works for me.

> any repro case?

Not sure if the cause of the leak is exactly the same, but simply starting/stopping a Xwayland app also leaks dmabuf: `parallel -j 10 "bash -c 'xclock & sleep 0.1 && kill \$!'" ::: $(seq 9999)` leaks 100 dmabuf per second on my machine

I bisected with this repro command and found 9994b73ad0af5c9ba5fc4058234cea77d0a1ddb1 to be the first bad commit

**Comment by vaxerski** [comment]
thanks, that helps a lot. It's likely that somehow the release events never get sent and thus xwayland thinks the buffers are still in use.

For future reference to myself. I am busy with #6608, after that I'll take a look here if it's not fixed yet by someone else.

My guess is some invisible or unmapped windows somehow skip the logic to send `sendRelease()` somewhere.

**Comment by gulafaran** [comment]
> Experienced same dmabuf leak issue with Xwayland Qt app (krita), it leaks ~500 dmabuf per hour and crashes Hyprland/Xwayland. The workaround mentioned in this issue works for me.
> 
> > any repro case?
> 
> Not sure if the cause of the leak is exactly the same, but simply starting/stopping a Xwayland app also leaks dmabuf: `parallel -j 10 "bash -c 'xclock & sleep 0.1 && kill \$!'" ::: $(seq 9999)` leaks 100 dmabuf per second on my machine
> 
> I bisected with this repro command and found [9994b73](https://github.com/hyprwm/Hyprland/commit/9994b73ad0af5c9ba5fc4058234cea77d0a1ddb1) to be the first bad commit

does this still happend with hyprutils/hyprwayland-scanner/hyprland built from git today? i cant seem to reproduce or get valgrind/asan to show any fd leaks

**Comment by sifmelcara** [comment]
> does this still happend with hyprutils/hyprwayland-scanner/hyprland built from git today? i cant seem to reproduce or get valgrind/asan to show any fd leaks

Yes, still reproducible. I build hyprland by running `nix build .?submodules=1#hyprland`, the version of hyprutils/hyprwayland-scanner can be found in `flake.lock`. Maybe try sleep more than 0.1 second to see if it helps? Can you repro with `v0.41.1` or `v0.41.2`?

You should be able to see a lot of dmabuf by running `ls -l /proc/$(pidof Hyprland)/fd`. Those fd are closed after XWayland get killed or stopped, so valgrind might not detect the issue here.

**Comment by gulafaran** [comment]
> > does this still happend with hyprutils/hyprwayland-scanner/hyprland built from git today? i cant seem to reproduce or get valgrind/asan to show any fd leaks
> 
> Yes, still reproducible. I build hyprland by running `nix build .?submodules=1#hyprland`, the version of hyprutils/hyprwayland-scanner can be found in `flake.lock`. Maybe try sleep more than 0.1 second to see if it helps? Can you repro with `v0.41.1` or `v0.41.2`?
> 
> You should be able to see a lot of dmabuf by running `ls -l /proc/$(pidof Hyprland)/fd`. Those fd are closed after XWayland get killed or stopped, so valgrind might not detect the issue here.

okay yeah reproduced it now, il see if i can figure it out

**Comment by vaxerski** [comment]
Thanks @gulafaran for looking into this, You can change the Xwayland command in src/xwayland/Server.cpp@295

if you add `WAYLAND_DEBUG=1 .........command......... > /home/amogus/log.txt` you should be able to inspect xwayland's wayland_debug. My suspicion is that some dmabufs are never `.release()`d, so xwayland thinks they are in use.

**Comment by gulafaran** [comment]
> Thanks @gulafaran for looking into this, You can change the Xwayland command in src/xwayland/Server.cpp@295
> 
> if you add `WAYLAND_DEBUG=1 .........command......... > /home/amogus/log.txt` you should be able to inspect xwayland's wayland_debug. My suspicion is that some dmabufs are never `.release()`d, so xwayland thinks they are in use.

yeah idk im giving it a rest, put breakpoints all over. ran it through valgrind gdb server tried tracking it down. changed all sorts of things to print all over im not finding the exact cause, its 100% reproduceable in a nested hyprland. just launch xclock and close it and it leaks dmabuf fds, `parallel -j 10 "bash -c 'xclock & sleep 1 && kill \$!'" ::: $(seq 9999)` run that for a few seconds and abort it. and you will see the fd's reaching large amounts fast.

**Comment by vaxerski** [comment]
is this still a thing after #6608 ?

**Comment by hdm9527** [comment]
> is this still a thing after #6608 ?

yes
```shell
❯ uptime
 19:45:52 up 10:18,  2 users,  load average: 1.62, 1.16, 1.05
❯ l /proc/$(pidof Hyprland)/fd | grep dmabuf | wc -l
5892
```

**Comment by vaxerski** [comment]
should be fixed, please check.

**Comment by hdm9527** [comment]
Thanks. It can be released correctly.

**Comment by hdm9527** [comment]
There's a problem with modifying cursor size memory not being released.
```shell
❯ grep VmRSS /proc/$(pidof Hyprland)/status
VmRSS:	1801040 kB
```
Steps to reproduce

1. `hyprctl setcursor Bibata-Modern-Ice 1000`
2. `hyprctl setcursor Bibata-Modern-Ice 24`
3. Allocated memory not released after modifying cursor.

**Comment by vaxerski** [comment]
that should be reported as a separate issue. Closing as this one is fixed
